aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +27,348 @@ namespace at {
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor3, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor4, 0)
+template <typename scalar1, typename scalar2, typename Op>
+AT_API inline void CPU_tensor_apply2(Tensor tensor1, Tensor tensor2, const Op op) {
+  if (!_apply_preamble({tensor1, tensor2}))
+    return;
+  if (tensor1.ndimension() < 8 && tensor2.ndimension() < 8) {
+    apply_op(
+        tensor1.numel(),
+        0,
+        op,
+        strided_tensor_iter_fixed<scalar1, 8>(tensor1),
+        strided_tensor_iter_fixed<scalar2, 8>(tensor2));
+  } else {
+    apply_op(
+        tensor1.numel(),
+        0,
+        op,
+        strided_tensor_iter<scalar1>(tensor1),
+        strided_tensor_iter<scalar2>(tensor2));
   }
-  if(tensor1_counter != NULL)
-    delete [] tensor1_counter;
-  if(tensor2_counter != NULL)
-    delete [] tensor2_counter;
-  if(tensor3_counter != NULL)
-    delete [] tensor3_counter;
-  if(tensor4_counter != NULL)
-    delete [] tensor4_counter;
 }
 
-/*
-  Apply a pointwise operator to four tensors.
-
-  The calling convention for op is a function/functor that takes takes four references to
-  type scalar; at least one of these references should be non-const in order to write the output.
-  For example, to compute a = b + c * d, op would be of the form:
-  [](scalar &a_val, const scalar &b_val, const scalar &c_val, const scalar &d_val) {
-    a_val = b_val + c_val * d_val;
-  };
-*/
-template<typename scalar1, typename scalar2, typename scalar3, typename scalar4, typename Op>
-void CPU_tensor_apply4(Tensor tensor1, Tensor tensor2, Tensor tensor3, Tensor tensor4, Op op) {
-  CPU_tensor_apply4_dim<scalar1, scalar2, scalar3, scalar4, Op>(tensor1, tensor2, tensor3, tensor4, -1, op);
+template <typename scalar1, typename scalar2, typename scalar3, typename Op>
+AT_API inline void CPU_tensor_apply3(
+    Tensor tensor1,
+    Tensor tensor2,
+    Tensor tensor3,
+    const Op op) {
+  if (!_apply_preamble({tensor1, tensor2, tensor3}))
+    return;
+  apply_op(
+      tensor1.numel(),
+      0,
+      op,
+      strided_tensor_iter<scalar1>(tensor1),
+      strided_tensor_iter<scalar2>(tensor2),
+      strided_tensor_iter<scalar3>(tensor3));
 }
 
+template <
+    typename scalar1,
+    typename scalar2,
+    typename scalar3,
+    typename scalar4,
+    typename Op>
+AT_API inline void CPU_tensor_apply4(
+    Tensor tensor1,
+    Tensor tensor2,
+    Tensor tensor3,
+    Tensor tensor4,
+    const Op op) {
+  if (!_apply_preamble({tensor1, tensor2, tensor3, tensor4}))
+    return;
+  apply_op(
+      tensor1.numel(),
+      0,
+      op,
+      strided_tensor_iter<scalar1>(tensor1),
+      strided_tensor_iter<scalar2>(tensor2),
+      strided_tensor_iter<scalar3>(tensor3),
+      strided_tensor_iter<scalar4>(tensor4));
 }
+
+template <typename scalar1, typename Op>
+AT_API inline void CPU_tensor_parallel_apply1(Tensor tensor1, const Op op) {
```

Comment:
When would one want to use `parallel_applyX` vs `applyX`? If something, I think we want this one to be the default, and we can always fall back to `serial_applyX` if someone is doing unsafe stuff.

Reply:
Sure, we can rename applyX to serial_applyX. I think it's important for this to be explicit, because some lambdas have state (such as a random number generator) that they want to capture by reference.